### DISCUSSION TITLE
Unsquashing from a merge commit is weird

### DISCUSSION THREAD

**Discussion by cormacrelf** [jj-vcs/jj#4343] [Q&A] [answered]
Say you've got a merge commit.

```sh
; jj new knx ttl
; jj log
@    qqyvktys me@example.com 2024-08-26 12:24:32 76ca4eb8
├─╮  (empty) (no description set)
│ ○  ttlomumr me@example.com 2024-08-26 11:29:21 ffbaeec4
│ │  add CONTRIBUTING.md
○ │  knxrzpxl me@example.com 2024-08-26 11:30:01 dfa6e45f
├─╯  add README.md
◆  zzzzzzzz root() 00000000
```

Then, you make some changes:
```sh
; echo 'fn main() {}' > main.rs
; jj log
@    qqyvktys me@example.com 2024-08-26 12:26:21 0f41bee4
├─╮  (no description set)
│ ○  ttlomumr me@example.com 2024-08-26 11:29:21 ffbaeec4
│ │  add CONTRIBUTING.md
○ │  knxrzpxl me@example.com 2024-08-26 11:30:01 dfa6e45f
├─╯  add README.md
◆  zzzzzzzz root() 00000000
```

Often at this point I realise I didn't want to do the work in the merge commit, since there were no conflicts to resolve or warnings/formatting issues to fix up during the merge, rather this was new work. As a habit it's better to `jj new` after a merge, but nevertheless.

There is no way to `jj split` or `jj commit` with all changes going into the second commit, from the command line. You have to open up a merge tool and manually move changes to the right. That job is basically what unsquash is for -- except this is what happens when you unsquash:

```sh
; jj new
; jj unsquash
; jj log
@    womzqlzp me@example.com 2024-08-26 12:32:55 a2592df4
├─╮  (no description set)
│ ○  ttlomumr me@example.com 2024-08-26 11:29:21 ffbaeec4
│ │  add CONTRIBUTING.md
○ │  knxrzpxl me@example.com 2024-08-26 11:30:01 dfa6e45f
├─╯  add README.md
◆  zzzzzzzz root() 00000000
```

Basically qqy was "empty" after we took all the changes, so it got abandoned, and we end up with a new merge commit that's exactly the same as before. The current behaviour is a bit surprising to me because I don't really think of the parents of a commit as being "changes" that you can move around. `jj log` shows an empty merge commit as empty, and it will get abandoned when you switch away from an empty @ merge commit (IIRC). So when unsquash says you can take the changes, I don't really expect the parents to be part of that.

Not sure what the right path forward is for this.  Maybe merge commits should not be empty for unsquash purposes? Or maybe there should be an `unsquash --allow-empty`?

Also, it would be nice if `jj split` could have a flag to invert the default location to the second commit. Don't know if invert is the right name, but:

```sh
; jj split --invert         # basically inserts an empty commit before rev, doesn't go into interactive mode
; jj split --invert PATHS   # puts PATHS in the first commit instead
```

**Comment by martinvonz** [comment, marked as answer]
I think you can `jj new --before qq` to insert a new commit before `qq`. It should become a merge commit because it inherits the parents of `qq` by default (you could combine it with `--after`, but that's not what you're looking for in your case).

EDIT: Add `--no-edit` to just create the commit without editing it afterwards.

**Reply by cormacrelf** [reply]
Hell yeah, that works. Thanks!
